rfd/0087-access-request-notification-routing.md

```diff
@@ -0,0 +1,207 @@
+
+The motivation behind adding a new resource, opposed to add new fields in `role` is to
+limit the `role` spec inflation and allow more flexibility to create role-agnostic routing rules.
+
+Routing rules are evaluated on access-request creation and will add elements to a new field in
+access-requests: `targets`.
+
+An example access_routing_rule would look like:
+
+```yaml
+kind: access_routing_rule
+version: v1
+metadata:
+  name: example
+spec:
+  targets:
+  # simplified syntax
+  - condition: 'equals(resource.spec.requested_role, "prod-rw")'
+    recipient: "Teleport Alice"
+    plugin: "pagerduty"
+  # raw expression
+  - expression: >
+      ifelse(
+        equals(resource.spec.requested_role, "prod-rw"),
+        pair(),
+        pair("msteams", "alice@example.com") 
+      )
+```
+
+Invoked with the role `"prod-rw"`, this would produce the following `access-request`
+
+```yaml
+kind: access_request
+spec:
+  user: bob
+  roles: prod-rw
+  state: 1
+  # [some fields were omitted for clarity]
+  request_reason: "Prod is down and rollback failed"
+  targets:
+    - plugin: pagerduty
+      recipient: Teleport Alice
+```
+
+From the plugin point of view, the recipients are additive:
+- they can come from the `role_to_recipient` map in the plugin configuration;
+- they can come from the `targets` on the access-request;
+- they can come from the requests suggested reviewers (under specific conditions); and
+- they can come from annotations (backward compatibility for pagerduty plugin).
+
+### Performance
+
+All `access_routing_rules` have to be evaluated against each access request. Even if this only applies
+on access request creation, this can be an intensive operation, especially for the database.
+Each auth server should maintain a cache of all target rules to avoid having to request them for each request.
+
+### Security
+
+In case of conflict (teamA writes a routing rule matching access-requests from teamB), both targets
+will be added to the access request. A misconfigured rule will result in a team being wrongly notified.
+The main issue will be spam/noise but this should not lock out the rightful team to get notified by
+the access request.
+
+Misconfigured rules should not represent security issues unless the plugin does some kind of automatic approval.
+This is the case for the pagerduty plugin which auto-approves if the requestor is on-call for the notified service.
+In such situation, `access_routing_rules` should be considered as critical as `roles` themselves as they can lead
+to automatic access escalation. Such risk is already present in the current implementation.
+
+This risk can be mitigated by not granting the access-request edition rights to the plugin (at the price of a feature loss).
+
+If the two targets are identical it is the plugin's responsibility to deduplicate before sending notifications
+(to avoid sending duplicates and to ensure `PluginData` consistence).
+
+### User Experience
```

Comment:
Let's give a specific example here of how a scenario from https://github.com/gravitational/teleport-plugins/issues/597 could be implemented using access request routing rules.

Reply:
Thanks for the feedback, this helped me identify the required predicate helpers.